crates/proto/src/rr/rdata/sshfp.rs

```diff
@@ -0,0 +1,220 @@
+
+/// [RFC 4255](https://tools.ietf.org/html/rfc4255#section-3.1)
+///
+/// ```text
+/// 3.1.  The SSHFP RDATA Format
+///
+///    The RDATA for a SSHFP RR consists of an algorithm number, fingerprint
+///    type and the fingerprint of the public host key.
+///
+///        1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 3 3
+///        0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+///        +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+///        |   algorithm   |    fp type    |                               /
+///        +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+                               /
+///        /                                                               /
+///        /                          fingerprint                          /
+///        /                                                               /
+///        +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+///
+/// 3.1.3.  Fingerprint
+///
+///    The fingerprint is calculated over the public key blob as described
+///    in [7].
+///
+///    The message-digest algorithm is presumed to produce an opaque octet
+///    string output, which is placed as-is in the RDATA fingerprint field.
+/// ```
+#[derive(Debug, PartialEq, Eq, Hash, Clone)]
+pub struct SSHFP {
+    algorithm: Algorithm,
+    fingerprint_type: FingerprintType,
+    fingerprint: Vec<u8>,
+}
+
+impl SSHFP {
+    /// Creates a new SSHFP record data.
+    ///
+    /// # Arguments
+    ///
+    /// * `algorithm` - the SSH public key algorithm.
+    /// * `fingerprint_type` - the fingerprint type to use.
+    /// * `fingerprint` - the fingerprint of the public key.
+    pub fn new(
+        algorithm: Algorithm,
+        fingerprint_type: FingerprintType,
+        fingerprint: Vec<u8>,
+    ) -> Self {
+        SSHFP {
+            algorithm,
+            fingerprint_type,
+            fingerprint,
+        }
+    }
+
+    /// The SSH public key algorithm.
+    pub fn algorithm(&self) -> Algorithm {
+        self.algorithm
+    }
+
+    /// The fingerprint type to use.
+    pub fn fingerprint_type(&self) -> FingerprintType {
+        self.fingerprint_type
+    }
+
+    /// The fingerprint of the public key.
+    pub fn fingerprint(&self) -> &[u8] {
+        &self.fingerprint
+    }
+}
+
+/// ```text
+/// 3.1.1.  Algorithm Number Specification
+///
+///    This algorithm number octet describes the algorithm of the public
+///    key.  The following values are assigned:
+///
+///           Value    Algorithm name
+///           -----    --------------
+///           0        reserved
+///           1        RSA
+///           2        DSS
+///
+///    Reserving other types requires IETF consensus [4].
+/// ```text
+///
+/// The algorithm values have been updated in
+/// [RFC 6594](https://tools.ietf.org/html/rfc6594) and
+/// [RFC 7479](https://tools.ietf.org/html/rfc7479).
+#[derive(Debug, PartialEq, Eq, Hash, Clone, Copy)]
+pub enum Algorithm {
```

Comment:
This competes with a similar name, if this generally only used in this context, it's probably ok, but an FYI: https://github.com/bluejekyll/trust-dns/blob/master/crates/proto/src/rr/dnssec/algorithm.rs#L104